llama/llama.go

```diff
@@ -315,18 +319,38 @@ func (m *Model) ApplyLoraFromFile(context *Context, loraPath string, scale float
 type Batch struct {
 	c         C.struct_llama_batch
 	batchSize int
+	maxSeq    int
 	embedSize int
 }
 
-// Creates a new batch for either word tokens if embed is 0 or
-// image embeddings if embed is specified. Batches cannot contain
-// both types at the same time
-func NewBatch(nTokens int, embed int, maxSeq int) *Batch {
-	return &Batch{
-		c:         C.llama_batch_init(C.int(nTokens), C.int(embed), C.int(maxSeq)),
-		batchSize: nTokens,
-		embedSize: embed,
+// Creates a new batch for either word tokens or image embeddings (if embedSize is non-zero).
+// Batches cannot contain both types at the same time. batchSize is the maximum number of entries
+// that can be added per sequence
+func NewBatch(batchSize int, maxSeq int, embedSize int) (*Batch, error) {
+	b := Batch{
+		c:         C.llama_batch_init(C.int(batchSize*maxSeq), C.int(embedSize), C.int(maxSeq)),
+		batchSize: batchSize,
+		maxSeq:    maxSeq,
+		embedSize: embedSize,
+	}
+
+	nilPointer := (embedSize == 0 && b.c.token == nil) || (embedSize != 0 && b.c.embd == nil) ||
```

Comment:
nit: we could probably check some of these before creating `b := Batch{ ...` which may also clean up the chained `||`s here

Reply:
All of these checks are for failed mallocs inside of llama_batch_init(). That was original problem that triggered the unsafe.Slice error - llama_batch_init() malloced some memory, which failed and returned a NULL pointer but wasn't check inside that function. Ideally, the checks would actually be in there. So we can't check anything before calling llama_batch_init(). Doing it after that and before batch creation just saves some `b.` but doesn't change the structure.

Looking at this, I guess I should call llama_batch_free() in the event of an error since we can have partial success. In reality it doesn't matter if we will immediately kill the process but I might add that on principle.

Reply:
Makes sense, thanks! 

> Ideally, the checks would actually be in there.

Yeah was thinking this too, would be a great future change